rust/parquet/src/arrow/levels.rs

```diff
@@ -0,0 +1,692 @@
+                            dbg!(&repetition);
+                            def_index += 1;
+                            (1..array_len).for_each(|_| {
+                                println!("* Index {} definition is {}, and repetition is {}. Current def: {}", def_index, definition[def_index], parent_rep, current_def_level);
+                                repetition.push(current_def_level); // was parent_rep + 1
+                                def_index += 1;
+                            });
+                        } else {
+                            (0..array_len).for_each(|_| {
+                                println!("* Index {} definition is {}, and repetition is {}. Current def: {}", def_index, definition[def_index], parent_rep, current_def_level);
+                                repetition.push(0); // TODO: should it be anything else?
+                                // TODO: use an append instead of pushes
+                                def_index += 1;
+                            });
+                        }
+                    }
+                    None => {
+                        println!("+ Index {} definition is {}, and repetition is 0. Current def: {}", def_index, definition[def_index], current_def_level);
+                        // if definition[def_index] == current_def_level {
+                            repetition.push(0);
+                            def_index += 1;
+                            (1..array_len).for_each(|_| {
+                                repetition.push(1); // TODO: is it always 0 and 1?
+                                def_index += 1;
+                            });
+                        // } else {
+                        //     (0..array_len).for_each(|_| {
+                        //         repetition.push(0); // TODO: should it be anything else?
+                        //                             // TODO: use an append instead of pushes
+                        //         def_index += 1;
+                        //     });
+                        // }
+                    }
+                }
+            }
+        });
+    });
+
+        let lev = LevelInfo {
+            definition,
+            repetition: if !has_repetition {
+                None
+            } else {
+                Some(repetition)
+            },
+            definition_mask,
+            array_mask,
+            array_offsets,
+            is_list: has_repetition,
+            max_definition: current_def_level,
+            is_nullable,
+        };
+
+        println!("done");
+
+        lev
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_calculate_array_levels_twitter_example() {
+        // based on the example at https://blog.twitter.com/engineering/en_us/a/2013/dremel-made-simple-with-parquet.html
```

Comment:
👍

Reply:
I've removed this for now, stashed them somewhere, as these are relevant for list tests.
So the `levels.rs` is currently indirectly tested by the roundtrip tests.